File: Python3/401 - 600/475 Heaters.py
# 108ms, 86.59%; 15.7MB, 6.82%
class Solution:
    def findRadius(self, houses: list, heaters: list) -> int:
        houses.sort()
        heaters.sort()
        res, i, size = 0, 1, len(heaters)
        heaters.append(heaters[-1])
        for pos in houses:
            while i < size and heaters[i] <= pos:
                i += 1
            res = max(res, min(abs(pos - heaters[i - 1]), abs(pos - heaters[i])))
        return res

File: 401 - 600/475 Heaters.cc
// 44ms, 100.00%; 11.8MB, 11.11%
#include <bits/stdc++.h>
using namespace std;

static int x = []() {
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);
    return 0;
}();

class Solution {
public:
    int findRadius(vector<int> &houses, vector<int> &heaters) {
        sort(houses.begin(), houses.end());
        sort(heaters.begin(), heaters.end());
        int res = 0;
        int i = 1, size = heaters.size();
        heaters.push_back(heaters.back());
        for (auto pos : houses) {
            while (i < size && heaters[i] <= pos) ++i;
            res = max(res, min(abs(heaters[i - 1] - pos), abs(heaters[i] - pos)));
        }
        return res;
    }
};
